hosted/preparation/install_test.go

```diff
@@ -0,0 +1,247 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e_test
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/rancher-sandbox/ele-testhelpers/kubectl"
+	"github.com/rancher-sandbox/ele-testhelpers/rancher"
+	"github.com/rancher-sandbox/ele-testhelpers/tools"
+)
+
+var _ = Describe("Provision k3s cluster with Rancher", Label("install"), func() {
+	// Create kubectl context
+	// Default timeout is too small, so New() cannot be used
+	k := &kubectl.Kubectl{
+		Namespace:    "",
+		PollTimeout:  tools.SetTimeout(300 * time.Second),
+		PollInterval: 500 * time.Millisecond,
+	}
+
+	It("Install upstream k3s cluster", func() {
+		if proxy != "" {
+			By("Run local squid proxy in docker", func() {
+				// Configure proxy before K3s installation if requested
+				GinkgoLogr.Info("Starting squid proxy")
+
+				cwd, _ := os.Getwd()
+				GinkgoLogr.Info("Current working directory: " + cwd)
+
+				out, err := exec.Command("docker", "run", "-d", "--rm", "--name", "squid_proxy",
+					"--volume", cwd+"/squid.conf:/etc/squid/squid.conf",
+					"-p", "3128:3128", "ubuntu/squid").CombinedOutput()
+				GinkgoWriter.Println(string(out))
+				Expect(err).To(Not(HaveOccurred()))
+			})
+
+			By("Configure proxy in /etc/default/k3s", func() {
+				k3sConfigPath := "/etc/default/k3s"
+				k3sConfig := fmt.Sprintf(`HTTP_PROXY=http://%s
+HTTPS_PROXY=http://%s
+NO_PROXY=127.0.0.0/8,10.0.0.0/8,cattle-system.svc,172.16.0.0/12,192.168.0.0/16,.svc,.cluster.local`, proxyHost, proxyHost)
+				// Write the proxy configuration to the file as root
+				out, err := exec.Command("sh", "-c", fmt.Sprintf("echo '%s' | sudo tee %s", k3sConfig, k3sConfigPath)).CombinedOutput()
+				GinkgoWriter.Println(string(out))
+				Expect(err).To(Not(HaveOccurred()))
+			})
+		}
+
+		By("Installing K3s", func() {
+			// Get K3s installation script
+			fileName := "k3s-install.sh"
+			Eventually(func() error {
+				return tools.GetFileFromURL("https://get.k3s.io", fileName, true)
+			}, tools.SetTimeout(2*time.Minute), 10*time.Second).ShouldNot(HaveOccurred())
+
+			// Set command and arguments
+			installCmd := exec.Command("sh", fileName)
+			// INSTALL_K3S_VERSION should be already set by the action or user's shell
+			installCmd.Env = append(os.Environ(), "INSTALL_K3S_EXEC=--disable metrics-server --write-kubeconfig-mode 644")
+
+			// Retry in case of (sporadic) failure...
+			count := 1
+			Eventually(func() error {
+				// Execute K3s installation
+				out, err := installCmd.CombinedOutput()
+				GinkgoWriter.Printf("K3s installation loop %d:\n%s\n", count, out)
+				count++
+				return err
+			}, tools.SetTimeout(2*time.Minute), 5*time.Second).Should(BeNil())
```

Comment:
nit:
Add a failure message to `Eventually`. Also add it to other eventually blocks if you go ahead with the suggestion.
```suggestion
			}, tools.SetTimeout(2*time.Minute), 5*time.Second).Should(BeNil(), "Timed out while waiting for k3s to install")
```

Reply:
AFAIK `Should` is not designed to return a custom message on failure :(